### PR TITLE
Conglomerate of PRs with tests

### PR DESCRIPTION
**Pull request by jrwesolo**
I was looking at this module and saw a few open pull requests that seem to have been abandoned by their submitters. I went ahead and implemented some of the changes, complete with tests, for the following PRs: #29, #43, #49.

**Highlights**
* Simplify the snmpd.conf template with some ruby magic
* Allow the community and network variables to be strings or arrays
* Change default groups to be empty, prevent DDoS
* Allow `service_config_perms` to be a class parameter
* Validate additional arrays in `init` class
* Add test coverage for new code as well as some old code

### CODE REVIEW

File: templates/snmpd.conf.erb [hunk 1]
```diff
@@ -14,64 +14,42 @@ agentaddress <%= @agentaddress.join(',') %>
...
+<%- [*@rw_community6].compact.each do |c| -%>
+  <%- [*@rw_network6].compact.each do |n| -%>
+rocommunity6 <%= c %> <%= n %>
+  <%- end -%>
+<%- end -%>
 
 # ------------------------------------------------------------------------------
 # VACM Configuration
 #       sec.name       source        community
-<% if @com2sec.any? -%>
```
Comment (jrwesolo) [inline templates/snmpd.conf.erb]: No longer needed since we validate that `com2sec` is an array in `init.rb`.

File: templates/snmpd.conf.erb [hunk 2]
```diff
@@ -14,64 +14,42 @@ agentaddress <%= @agentaddress.join(',') %>
...
-<% @com2sec.each do |com2sec| -%>
-com2sec <%= com2sec %>
-<% end -%>
+<% @com2sec.each do |c| -%>
+com2sec <%= c %>
 <% end -%>
 
-<% if @com2sec6.any? -%>
```
Comment (jrwesolo) [inline templates/snmpd.conf.erb]: No longer needed since we validate that `com2sec6` is an array in `init.rb`.

File: templates/snmpd.conf.erb [hunk 3]
```diff
@@ -14,64 +14,42 @@ agentaddress <%= @agentaddress.join(',') %>
...
-<% @com2sec6.each do |com2sec6| -%>
-com2sec6 <%= com2sec6 %>
-<% end -%>
+<% @com2sec6.each do |c| -%>
+com2sec6 <%= c %>
 <% end -%>
 
 #       groupName      securityModel securityName
-<% if @groups.any? -%>
```
Comment (jrwesolo) [inline templates/snmpd.conf.erb]: No longer needed since we validate that `groups` is an array in `init.rb`.

File: templates/snmpd.conf.erb [hunk 4]
```diff
@@ -14,64 +14,42 @@ agentaddress <%= @agentaddress.join(',') %>
...
-<% if @rw_community6 and (@rw_community6.size > 0) -%>
-<% if (@rw_network6.is_a?(Array)) and (@rw_network6.count > 0) -%>
-<% @rw_network6.each do |rw_net6| -%>
-rwcommunity6 <%= @rw_community6 %> <%= rw_net6 %>
-<% end -%>
-<% else -%>
-rwcommunity6 <%= @rw_community6 %> <%= @rw_network6 %>
-<% end -%>
-<% end -%>
+<%- [*@ro_community].compact.each do |c| -%>
```
Comment (jrwesolo) [inline templates/snmpd.conf.erb]: The [*@ro_community] syntax will take a variable and convert it into an array if it is not already. This allows us to really trim down this file.

**Comment by razorsedge** [comment]
Awesome work @jrwesolo .  If I could ask one favour: could you resubmit against the `develop` branch?

**Comment by jrwesolo** [comment]
Sure thing! https://github.com/razorsedge/puppet-snmp/pull/62.

**Pull request closed without merging**